File: 1/2021.12.27/2122.Recover the Original Array/python/solution.py
from collections import defaultdict
class Solution:
    def recoverArray(self, nums):
        n = len(nums)
        nums.sort()
        for i in range(1, n):
            if nums[i] == nums[0] or (nums[i] - nums[0]) % 2 != 0:
                continue
            used = [False] * n
            used[0] = used[i] = True
            k = (nums[i] - nums[0]) // 2
            ans = [nums[0] + k]

            left, right = 0, i
            for j in range(1, n // 2):
                while used[left]:
                    left += 1
                while right < n and (used[right] or nums[right] - nums[left] != k * 2):
                    right += 1
                if right == n:
                    break
                ans.append(nums[left] + k)
                used[left] = used[right] = True

            if len(ans) == n // 2:
                return ans



def main():
    so = Solution()
    nums = [2, 10, 6, 4, 8, 12]
    print(so.recoverArray(nums))


if __name__ == '__main__':
    main()

File: 1/2021.12.27/2122.Recover the Original Array/java/solution.java
class Solution {

    public int[] recoverArray(int[] nums) {
        int n = nums.length;
        Arrays.sort(nums);
        for (int i = 1; i < n; ++i) {
            if (nums[i] == nums[0] || (nums[i] - nums[0]) % 2 != 0) {
                continue;
            }
            boolean[] used = new boolean[n];
            int[] ans = new int[n / 2];
            int k = (nums[i] - nums[0]) / 2;
            ans[0] = nums[0] + k;
            used[0] = used[i] = true;
            int left = 0, right = i;
            for (int j = 1; j < n / 2; ++j) {
                while (used[left]) {
                    ++left;
                }
                while (right < n && (used[right] || (nums[right] - nums[left] != k * 2))) {
                    ++right;
                }
                if (right == n) {
                    break;
                }
                ans[j] = nums[left] + k;
                used[left] = used[right] = true;
            }
            if (right != n) {
                return ans;
            }
        }

        // can not reach here
        return new int[0];
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        int[] nums = new int[]{2, 10, 6, 4, 8, 12};
        System.out.println(Arrays.toString(solution.recoverArray(nums)));
        nums = new int[]{5, 435};
        System.out.println(Arrays.toString(solution.recoverArray(nums)));
    }
}
